front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/PipelineTemplateController.java

```diff
@@ -0,0 +1,113 @@
+
+import com.netflix.spinnaker.front50.exception.NotFoundException;
+import com.netflix.spinnaker.front50.model.pipeline.PipelineTemplate;
+import com.netflix.spinnaker.front50.model.pipeline.PipelineTemplateDAO;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@RestController
+@RequestMapping("pipelineTemplates")
+public class PipelineTemplateController {
+
+  @Autowired
+  PipelineTemplateDAO pipelineTemplateDAO;
+
+  // TODO rz - Add fiat authz
+
+  @RequestMapping(value = "", method = RequestMethod.GET)
+  List<PipelineTemplate> list(@RequestParam(required = false, value = "scope", defaultValue = "global") List<String> scope) {
+    return (List<PipelineTemplate>) pipelineTemplateDAO.getPipelineTemplatesByScope(scope);
+  }
+
+  @RequestMapping(value = "", method = RequestMethod.POST)
+  void save(@RequestBody PipelineTemplate pipelineTemplate) {
+    checkForDuplicatePipelineTemplate(pipelineTemplate.getId());
+    pipelineTemplateDAO.create(pipelineTemplate.getId(), pipelineTemplate);
+  }
+
+  @RequestMapping(value = "{id}", method = RequestMethod.GET)
+  PipelineTemplate get(@PathVariable String id) {
+    return pipelineTemplateDAO.findById(id);
+  }
+
+  @RequestMapping(value = "{id}", method = RequestMethod.PUT)
+  PipelineTemplate update(@PathVariable String id, @RequestBody PipelineTemplate pipelineTemplate) {
+    PipelineTemplate existingPipelineTemplate = pipelineTemplateDAO.findById(id);
+    if (!pipelineTemplate.getId().equals(existingPipelineTemplate.getId())) {
+      throw new InvalidPipelineTemplateRequestException("The provided id " + id + " doesn't match the pipeline template id " + pipelineTemplate.getId());
+    }
+
+    pipelineTemplate.setLastModified(System.currentTimeMillis());
+    pipelineTemplateDAO.update(id, pipelineTemplate);
+
+    return pipelineTemplate;
+  }
+
+  private void checkForDuplicatePipelineTemplate(String id) {
+    try {
+      pipelineTemplateDAO.findById(id);
+    } catch (NotFoundException e) {
+      return;
+    }
+    throw new DuplicatePipelineTemplateIdException("A pipeline template with the id " + id + " already exists");
+  }
+
+  @ExceptionHandler(InvalidPipelineTemplateRequestException.class)
+  @ResponseStatus(HttpStatus.BAD_REQUEST)
+  Map handleInvalidPipelineTemplateRequestException(InvalidPipelineTemplateRequestException e) {
```

Comment:
Nothing to do with this PR but I reckon we can do a much better job of exception handling that doesn't require per-controller exception classes ...